crates/services/p2p/src/peer_report.rs

```diff
@@ -94,24 +109,55 @@ impl Behaviour {
             connected_reserved_nodes: Default::default(),
             pending_connections: Default::default(),
             pending_events: VecDeque::default(),
+            ips_to_peers: Default::default(),
+            max_connections_per_ip: config.max_peers_per_remote_ip,
             decay_interval: time::interval(Duration::from_secs(
                 REPUTATION_DECAY_INTERVAL_IN_SECONDS,
             )),
         }
     }
 }
 
+fn get_ip_addr(addr: &Multiaddr) -> Option<IpAddr> {
+    addr.iter().find_map(|p| match p {
+        Ip4(addr) => Some(IpAddr::V4(addr)),
+        Ip6(addr) => Some(IpAddr::V6(addr)),
+        _ => None,
+    })
+}
+
 impl NetworkBehaviour for Behaviour {
     type ConnectionHandler = dummy::ConnectionHandler;
     type ToSwarm = PeerReportEvent;
 
     fn handle_established_inbound_connection(
         &mut self,
         _connection_id: ConnectionId,
-        _peer: PeerId,
+        peer: PeerId,
         _local_addr: &Multiaddr,
-        _remote_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
     ) -> Result<THandler<Self>, ConnectionDenied> {
+        if let Some(ip) = get_ip_addr(remote_addr) {
```

Comment:
Is it possible that `get_ip_addr` could return two different IP for an IPv4 and his IPv6 mapped or libp2p prevents IPv4-mapped in IPv6 ?

If it's the first case then they will fall in two separate entries and in this case we might want `get_ip_addr` to canonicalise the IP (https://doc.rust-lang.org/std/net/enum.IpAddr.html#method.to_canonical)